File: src/pipyter/protocol/pigent.py
"""Pigent v0.2 protocol models.

Mirror of ``packages/protocol/src/pigent.ts`` and the authoritative JSON
Schemas in ``packages/protocol/schemas/pigent-*.schema.json`` (the schemas are
force-included in the wheel/sdist). Every public enum is a strict ``Literal``:
unknown tool names, actions, statuses, modes and protocol versions are
rejected, and legacy ``pilot`` mode maps to ``auto`` exactly once.
"""

from __future__ import annotations

from typing import Annotated, Any, Literal, TypeAlias

from pydantic import BaseModel, Field, StringConstraints

PIGENT_PROTOCOL_VERSION = "0.2"

# ---------------------------------------------------------------------------
# Enums (keep in sync with the JSON schemas and pigent.ts)
# ---------------------------------------------------------------------------

PigentToolName: TypeAlias = Literal[
    "read", "view", "write", "update", "bash", "notebook", "kernel", "inspect", "tasks", "delegate"
]
PIGENT_TOOL_NAMES: tuple[PigentToolName, ...] = (
    "read", "view", "write", "update", "bash", "notebook", "kernel", "inspect", "tasks", "delegate",
)

PigentMode: TypeAlias = Literal["ask", "plan", "auto"]
PIGENT_MODES: tuple[PigentMode, ...] = ("ask", "plan", "auto")

LegacyPigentMode: TypeAlias = Literal["pilot", "ask", "plan", "auto"]

PigentErrorCode: TypeAlias = Literal[
    "invalid_request",
    "invalid_path",
    "permission_denied",
    "not_found",
    "unsupported_media",
    "too_large",
    "revision_conflict",
    "mode_denied",
    "confirmation_required",
    "kernel_unavailable",
    "kernel_busy",
    "execution_timeout",
    "cancelled",
    "internal_error",
    "document_dirty",
    "output_persist_conflict",
    "kernel_dead",
    "model_configuration_required",
    "service_unavailable",
    "payload_missing",
    "payload_stale",
    "uv_missing",
    "uv_incompatible",
    "config_migration_conflict",
    "config_migration_invalid_source",
    "kernel_environment_not_found",
    "kernel_environment_conflict",
    "kernel_environment_busy",
    "kernel_environment_stale",
    "kernel_environment_provision_failed",
    "kernel_environment_sync_failed",
    "kernel_queue_cancelled",
    "operation_not_cancellable",
    "interaction_superseded",
]
PIGENT_ERROR_CODES: tuple[PigentErrorCode, ...] = (
    "invalid_request", "invalid_path", "permission_denied", "not_found", "unsupported_media",
    "too_large", "revision_conflict", "mode_denied", "confirmation_required",
    "kernel_unavailable", "kernel_busy", "execution_timeout", "cancelled", "internal_error",
    "document_dirty", "output_persist_conflict", "kernel_dead",
    "model_configuration_required", "service_unavailable", "payload_missing", "payload_stale",
    "uv_missing", "uv_incompatible", "config_migration_conflict", "config_migration_invalid_source",
    "kernel_environment_not_found", "kernel_environment_conflict", "kernel_environment_busy",
    "kernel_environment_stale", "kernel_environment_provision_failed", "kernel_environment_sync_failed",
    "kernel_queue_cancelled", "operation_not_cancellable", "interaction_superseded",
)

PigentEventType: TypeAlias = Literal[
    "session.created", "session.updated", "mode.changed", "assistant.text", "assistant.thinking",
    "tool.start", "tool.update", "tool.end", "tasks.snapshot", "delegate.start",
    "delegate.update", "delegate.end", "interaction.required", "interaction.resolved",
    "context.updated", "kernel.updated", "artifact.created", "error", "aborted", "settled",
    "reconnect.cursor", "operation.started", "operation.updated", "operation.ended",
    "kernel.environment.updated",
]
PIGENT_EVENT_TYPES: tuple[PigentEventType, ...] = (
    "session.created", "session.updated", "mode.changed", "assistant.text", "assistant.thinking",
    "tool.start", "tool.update", "tool.end", "tasks.snapshot", "delegate.start",
    "delegate.update", "delegate.end", "interaction.required", "interaction.resolved",
    "context.updated", "kernel.updated", "artifact.created", "error", "aborted", "settled",
    "reconnect.cursor", "operation.started", "operation.updated", "operation.ended",
    "kernel.environment.updated",
)

PigentTaskStatus: TypeAlias = Literal["pending", "running", "done", "blocked", "failed"]
PIGENT_TASK_STATUSES: tuple[PigentTaskStatus, ...] = ("pending", "running", "done", "blocked", "failed")

PigentSessionStatus: TypeAlias = Literal["active", "completed", "failed", "interrupted", "waiting_for_user"]
PIGENT_SESSION_STATUSES: tuple[PigentSessionStatus, ...] = (
    "active", "completed", "failed", "interrupted", "waiting_for_user",
)

PigentArtifactKind: TypeAlias = Literal["image", "table", "text", "file"]
PIGENT_ARTIFACT_KINDS: tuple[PigentArtifactKind, ...] = ("image", "table", "text", "file")

PigentDelegateProfile: TypeAlias = Literal["analysis", "research", "review", "implementation"]
PIGENT_DELEGATE_PROFILES: tuple[PigentDelegateProfile, ...] = (
    "analysis", "research", "review", "implementation",
)

PigentCapability: TypeAlias = Literal[
    "filesystem.read", "filesystem.write", "visual.read", "notebook.read", "notebook.write",
    "kernel.status", "kernel.inspect", "kernel.execute", "process.execute",
    "process.interactive", "network", "system.execute", "tasks.write", "delegate.read",
    "delegate.write", "kernel.environment.read", "kernel.environment.manage",
]
PIGENT_CAPABILITIES: tuple[PigentCapability, ...] = (
    "filesystem.read", "filesystem.write", "visual.read", "notebook.read", "notebook.write",
    "kernel.status", "kernel.inspect", "kernel.execute", "process.execute",
    "process.interactive", "network", "system.execute", "tasks.write", "delegate.read",
    "delegate.write", "kernel.environment.read", "kernel.environment.manage",
)

CapabilityLevel: TypeAlias = Literal["allow", "deny", "os", "interactive"]

NOTBOOK_ACTIONS: tuple[str, ...] = (
    "read_cell", "update_cell", "insert_cell", "delete_cell", "move_cell", "run_cell",
    "add_markdown", "clear_output",
)
KERNEL_READ_ACTIONS: tuple[str, ...] = ("status", "list_environments", "operation_status")
KERNEL_ACTIONS: tuple[str, ...] = (
    "status", "execute", "interrupt", "restart", "shutdown", "list_environments", "operation_status",
    "create_temporary", "create_maintained", "sync_environment", "start_environment",
    "promote_environment", "delete_environment",
)
INSPECT_ACTIONS: tuple[str, ...] = ("variables", "variable", "dataframe", "figure", "object")
TASKS_ACTIONS: tuple[str, ...] = ("get", "replace", "patch")

# Frozen mode x capability matrix (03-modes-permissions.md). Ask and Plan
# restrictions define product intent; Auto is the single execution mode and is
# never degraded into a hidden restricted tier.
PIGENT_MODE_MATRIX: dict[PigentCapability, dict[PigentMode, CapabilityLevel]] = {
    "filesystem.read": {"ask": "allow", "plan": "allow", "auto": "os"},
    "visual.read": {"ask": "allow", "plan": "allow", "auto": "os"},
    "notebook.read": {"ask": "allow", "plan": "allow", "auto": "os"},
    "kernel.status": {"ask": "allow", "plan": "allow", "auto": "os"},
    "kernel.inspect": {"ask": "allow", "plan": "allow", "auto": "os"},
    "kernel.environment.read": {"ask": "allow", "plan": "allow", "auto": "allow"},
    "tasks.write": {"ask": "deny", "plan": "allow", "auto": "allow"},
    "delegate.read": {"ask": "allow", "plan": "allow", "auto": "allow"},
    "filesystem.write": {"ask": "deny", "plan": "deny", "auto": "os"},
    "notebook.write": {"ask": "deny", "plan": "deny", "auto": "os"},
    "kernel.execute": {"ask": "deny", "plan": "deny", "auto": "os"},
    "process.execute": {"ask": "deny", "plan": "deny", "auto": "os"},
    "process.interactive": {"ask": "deny", "plan": "deny", "auto": "interactive"},
    "network": {"ask": "deny", "plan": "deny", "auto": "os"},
    "system.execute": {"ask": "deny", "plan": "deny", "auto": "interactive"},
    "delegate.write": {"ask": "deny", "plan": "deny", "auto": "allow"},
    "kernel.environment.manage": {"ask": "deny", "plan": "deny", "auto": "os"},
}

# Projected tool catalogs by mode.
PIGENT_CATALOGS: dict[PigentMode, tuple[PigentToolName, ...]] = {
    "ask": ("read", "view", "notebook", "kernel", "inspect", "delegate"),
    "plan": ("read", "view", "notebook", "kernel", "inspect", "delegate", "tasks"),
    "auto": PIGENT_TOOL_NAMES,
}

# Per-mode action filters for multi-action tools.
PIGENT_ACTION_FILTERS: dict[str, dict[PigentMode, tuple[str, ...]]] = {
    "notebook": {
        "ask": ("read_cell",),
        "plan": ("read_cell",),
        "auto": NOTBOOK_ACTIONS,
    },
    "kernel": {
        "ask": KERNEL_READ_ACTIONS,
        "plan": KERNEL_READ_ACTIONS,
        "auto": KERNEL_ACTIONS,
    },
    "inspect": {
        "ask": INSPECT_ACTIONS,
        "plan": INSPECT_ACTIONS,
        "auto": INSPECT_ACTIONS,
    },
    "tasks": {
        "ask": (),
        "plan": TASKS_ACTIONS,
        "auto": TASKS_ACTIONS,
    },
    "delegate": {
        "ask": ("analysis", "research", "review"),
        "plan": ("analysis", "research", "review"),
        "auto": PIGENT_DELEGATE_PROFILES,
    },
}


def allowed_tools(mode: PigentMode) -> tuple[PigentToolName, ...]:
    """Tools advertised to the model in the selected mode."""
    if mode not in PIGENT_CATALOGS:
        raise ValueError(f"unknown Pigent mode: {mode!r}")
    return PIGENT_CATALOGS[mode]


def allowed_actions(tool: str, mode: PigentMode) -> tuple[str, ...]:
    """Actions allowed for a (multi-action) tool in the selected mode."""
    filter_ = PIGENT_ACTION_FILTERS.get(tool)
    if filter_ is None:
        return ()
    return filter_[mode]


def migrate_legacy_mode(value: Any) -> PigentMode:
    """Map legacy pilot state to auto; reject anything else unknown."""
    if value == "pilot":
        return "auto"
    if value in PIGENT_MODES:
        return value  # type: ignore[return-value]
    raise ValueError(f"invalid mode: {value!r} (pilot maps to auto, new schemas reject pilot)")


def migrate_legacy_session_state(state: dict[str, Any]) -> tuple[dict[str, Any], bool]:
    """Migrate legacy session state: mode/requested_mode/effective_mode pilot -> auto.

    Legacy fields are removed after one successful write in the new schema.
    Returns (migrated, changed).
    """
    migrated = dict(state)
    changed = False
    for key in ("mode", "requested_mode", "effective_mode"):
        if key in migrated:
            value = migrated[key]
            if value == "pilot":
                migrated[key] = "auto"
                changed = True
            elif value not in PIGENT_MODES:
                raise ValueError(f"invalid legacy mode field {key}: {value!r}")
    if changed:
        migrated.pop("requested_mode", None)
        migrated.pop("effective_mode", None)
    return migrated, changed


# ---------------------------------------------------------------------------
# Pydantic models
# ---------------------------------------------------------------------------

Revision = Annotated[str, StringConstraints(pattern=r"^sha256:[0-9a-f]{64}$")]
"""Opaque content revision: sha256:<64 hex>."""


def _revision_pattern() -> str:
    return r"^sha256:[0-9a-f]{64}$"


class ArtifactRef(BaseModel):
    id: str
    kind: PigentArtifactKind
    mime: str
    size: int = Field(ge=0)
    created_at: str
    hash: str = Field(pattern=_revision_pattern())
    width: int | None = Field(default=None, ge=0)
    height: int | None = Field(default=None, ge=0)
    expires_at: str | None = None


class PigentToolError(BaseModel):
    code: PigentErrorCode
    message: str
    retryable: bool = False
    details: dict[str, Any] = Field(default_factory=dict)


class Revisions(BaseModel):
    """Structured before/after revision pair returned by mutating tools."""

    before: Revision
    after: Revision


class PigentToolResult(BaseModel):
    """Versioned envelope returned by every Python-backed tool."""

    version: Literal[1] = 1
    ok: bool
    summary: str = ""
    data: dict[str, Any] = Field(default_factory=dict)
    artifacts: list[ArtifactRef] = Field(default_factory=list)
    revisions: Revisions | None = Field(default=None)
    error: PigentToolError | None = None
    warnings: list[str] = Field(default_factory=list)


class PigentToolContext(BaseModel):
    """Trusted context injected by the Pigent host; the model never overrides it."""

    protocol_version: Literal["0.2"] = "0.2"
    tool_call_id: str
    session_id: str
    workspace_id: str
    mode: PigentMode
    active_document: dict[str, str] | None = None
    active_kernel_id: str | None = None


OperationState: TypeAlias = Literal["queued", "running", "waiting_for_user", "succeeded", "failed", "cancelled"]
OperationOutcome: TypeAlias = Literal["success", "partial", "failed", "cancelled", "superseded"]
KernelEnvironmentKind: TypeAlias = Literal["temporary", "maintained"]
KernelEnvironmentStatus: TypeAlias = Literal["provisioning", "ready", "stale", "syncing", "error", "deleting", "missing"]


class OperationProgress(BaseModel):
    phase: str
    completed: int = Field(default=0, ge=0)
    total: int | None = Field(default=None, ge=0)
    message: str = ""


class ToolReceipt(BaseModel):
    outcome: OperationOutcome
    summary: str
    identifiers: dict[str, str] = Field(default_factory=dict)
    at: str


class OperationResource(BaseModel):
    type: Literal["kernel_environment"]
    id: str


class OperationEnvelope(BaseModel):
    operation_id: str
    kind: str
    state: OperationState
    progress: OperationProgress | None = None
    resource: OperationResource
    created_at: str
    updated_at: str
    session_id: str | None = None
    tool_call_id: str | None = None
    cancellable: bool = True
    receipt: ToolReceipt | None = None
    error: PigentToolError | None = None


class KernelEnvironmentSummary(BaseModel):
    id: str
    kind: KernelEnvironmentKind
    name: str | None = None
    display_name: str
    status: KernelEnvironmentStatus
    python_request: str
    python_version: str | None = None
    interpreter: str | None = None
    packages: list[str] = Field(default_factory=list)
    source: dict[str, Any] | None = None
    lock_revision: str | None = None
    revision: str
    created_at: str
    updated_at: str
    last_used_at: str | None = None
    expires_at: str | None = None
    active_kernel_ids: list[str] = Field(default_factory=list)
    last_error: PigentToolError | None = None


class ExecutionIdentity(BaseModel):
    username: str
    uid: int | str | None = None
    home: str
    workspace: str


class ModeState(BaseModel):
    """Persisted mode selection. No requested/effective split."""

    mode: PigentMode
    changed_by: Literal["user", "host"] = "user"
    changed_at: str
    approval_preference: Literal["automatic", "review_all"] = "automatic"


class TaskNode(BaseModel):
    id: str
    title: str
    status: PigentTaskStatus
    depends_on: list[str] = Field(default_factory=list)
    completion_criteria: list[str] = Field(default_factory=list)
    children: list[TaskNode] = Field(default_factory=list)


class TasksSnapshot(BaseModel):
    revision: str
    root: TaskNode
    updated_at: str | None = None


class PigentSession(BaseModel):
    id: str
    account_id: str
    project_id: str
    workspace_id: str
    node_id: str
    mode: PigentMode
    approval_preference: Literal["automatic", "review_all"] = "automatic"
    status: PigentSessionStatus
    title: str | None = None
    created_at: str
    last_activity_at: str
    active_document: dict[str, str] | None = None
    active_kernel_id: str | None = None
    model: dict[str, str] | None = None
    tasks_snapshot: TasksSnapshot | None = None


class TerminalSession(BaseModel):
    """A persistent human Shell/PTY session; one tab is one real process."""

    id: str
    name: str
    executable: str
    cwd: str
    status: Literal["running", "exited", "closed"]
    cols: int = Field(default=80, ge=2, le=500)
    rows: int = Field(default=24, ge=1, le=500)
    created_at: str
    last_exit_code: int | None = None


class PigentInteraction(BaseModel):
    """Human interaction is a host event, never an LLM tool."""

    version: Literal[1] = 1
    interaction_id: str
    session_id: str
    tool_call_id: str
    kind: Literal["pty_handoff", "review_request", "clarification"]
    summary: str
    shell_session_id: str | None = None
    command_preview: str | None = None
    choices: list[Literal["open_shell", "cancel", "allow_once", "allow_workspace", "reject"]] = Field(
        min_length=1
    )


class PigentEvent(BaseModel):
    """Public browser event; reconnect.cursor has no business event ID."""

    version: Literal[1] = 1
    event_id: int | None = Field(default=None, ge=1)
    session_id: str
    type: PigentEventType
    timestamp: str
    payload: dict[str, Any] = Field(default_factory=dict)


class WorkspaceContext(BaseModel):
    """Active workspace context published independently of prompts."""

    type: Literal["workspace.context"] = "workspace.context"
    active_document: str
    document_revision: str | None = None
    active_cell_id: str | None = None
    selection: str | None = None
    active_kernel_id: str | None = None
    selected_figure_id: str | None = None

File: packages/protocol/src/pigent.ts
/**
 * Pigent v0.2 contracts (Python equivalents: src/pipyter/protocol/pigent.py).
 * Authoritative JSON Schema lives in packages/protocol/schemas/pigent-*.schema.json;
 * the wheel force-includes those schemas.
 *
 * Product surface (docs/plans/pigent-v0.1/):
 *  - exactly ten LLM-visible tools: read view write update bash notebook kernel inspect tasks delegate
 *  - exactly three modes: Ask, Plan, Auto. Pilot is rejected by every new schema;
 *    legacy state maps to auto exactly once.
 *  - Auto executes with the same practical authority as the runtime OS user; no
 *    workspace-only sandbox, denylist, or hidden restricted tier.
 */

export const PIGENT_PROTOCOL_VERSION = '0.2' as const

export const PIGENT_TOOL_NAMES = [
  'read',
  'view',
  'write',
  'update',
  'bash',
  'notebook',
  'kernel',
  'inspect',
  'tasks',
  'delegate',
] as const
export type PigentToolName = (typeof PIGENT_TOOL_NAMES)[number]

export const PIGENT_MODES = ['ask', 'plan', 'auto'] as const
export type PigentMode = (typeof PIGENT_MODES)[number]

/** Legacy pilot values accepted only by the migration loader. */
export type LegacyPigentMode = PigentMode | 'pilot'

export const PIGENT_ERROR_CODES = [
  'invalid_request',
  'invalid_path',
  'permission_denied',
  'not_found',
  'unsupported_media',
  'too_large',
  'revision_conflict',
  'mode_denied',
  'confirmation_required',
  'kernel_unavailable',
  'kernel_busy',
  'execution_timeout',
  'cancelled',
  'internal_error',
  'document_dirty',
  'output_persist_conflict',
  'kernel_dead',
  'model_configuration_required',
  'service_unavailable',
  'payload_missing', 'payload_stale', 'uv_missing', 'uv_incompatible',
  'config_migration_conflict', 'config_migration_invalid_source',
  'kernel_environment_not_found', 'kernel_environment_conflict', 'kernel_environment_busy',
  'kernel_environment_stale', 'kernel_environment_provision_failed', 'kernel_environment_sync_failed',
  'kernel_queue_cancelled', 'operation_not_cancellable', 'interaction_superseded',
] as const
export type PigentErrorCode = (typeof PIGENT_ERROR_CODES)[number]

export const PIGENT_EVENT_TYPES = [
  'session.created',
  'session.updated',
  'mode.changed',
  'assistant.text',
  'assistant.thinking',
  'tool.start',
  'tool.update',
  'tool.end',
  'tasks.snapshot',
  'delegate.start',
  'delegate.update',
  'delegate.end',
  'interaction.required',
  'interaction.resolved',
  'context.updated',
  'kernel.updated',
  'artifact.created',
  'error',
  'aborted',
  'settled',
  'reconnect.cursor',
  'operation.started', 'operation.updated', 'operation.ended', 'kernel.environment.updated',
] as const
export type PigentEventType = (typeof PIGENT_EVENT_TYPES)[number]

export const PIGENT_TASK_STATUSES = ['pending', 'running', 'done', 'blocked', 'failed'] as const
export type PigentTaskStatus = (typeof PIGENT_TASK_STATUSES)[number]

export const PIGENT_SESSION_STATUSES = [
  'active',
  'completed',
  'failed',
  'interrupted',
  'waiting_for_user',
] as const
export type PigentSessionStatus = (typeof PIGENT_SESSION_STATUSES)[number]

export const PIGENT_ARTIFACT_KINDS = ['image', 'table', 'text', 'file'] as const
export type PigentArtifactKind = (typeof PIGENT_ARTIFACT_KINDS)[number]

export const PIGENT_DELEGATE_PROFILES = ['analysis', 'research', 'review', 'implementation'] as const
export type PigentDelegateProfile = (typeof PIGENT_DELEGATE_PROFILES)[number]

export const PIGENT_CAPABILITIES = [
  'filesystem.read',
  'filesystem.write',
  'visual.read',
  'notebook.read',
  'notebook.write',
  'kernel.status',
  'kernel.inspect',
  'kernel.execute',
  'process.execute',
  'process.interactive',
  'network',
  'system.execute',
  'tasks.write',
  'delegate.read',
  'delegate.write',
  'kernel.environment.read', 'kernel.environment.manage',
] as const
export type PigentCapability = (typeof PIGENT_CAPABILITIES)[number]

/** allow: available; deny: contradicts the selected mode; os: Auto + OS identity; interactive: Auto, may pause for direct input. */
export type CapabilityLevel = 'allow' | 'deny' | 'os' | 'interactive'

/** Frozen mode x capability matrix (03-modes-permissions.md). */
export const PIGENT_MODE_MATRIX: Readonly<Record<PigentCapability, Readonly<Record<PigentMode, CapabilityLevel>>>> = {
  'filesystem.read': { ask: 'allow', plan: 'allow', auto: 'os' },
  'visual.read': { ask: 'allow', plan: 'allow', auto: 'os' },
  'notebook.read': { ask: 'allow', plan: 'allow', auto: 'os' },
  'kernel.status': { ask: 'allow', plan: 'allow', auto: 'os' },
  'kernel.inspect': { ask: 'allow', plan: 'allow', auto: 'os' },
  'kernel.environment.read': { ask: 'allow', plan: 'allow', auto: 'allow' },
  'tasks.write': { ask: 'deny', plan: 'allow', auto: 'allow' },
  'delegate.read': { ask: 'allow', plan: 'allow', auto: 'allow' },
  'filesystem.write': { ask: 'deny', plan: 'deny', auto: 'os' },
  'notebook.write': { ask: 'deny', plan: 'deny', auto: 'os' },
  'kernel.execute': { ask: 'deny', plan: 'deny', auto: 'os' },
  'process.execute': { ask: 'deny', plan: 'deny', auto: 'os' },
  'process.interactive': { ask: 'deny', plan: 'deny', auto: 'interactive' },
  network: { ask: 'deny', plan: 'deny', auto: 'os' },
  'system.execute': { ask: 'deny', plan: 'deny', auto: 'interactive' },
  'delegate.write': { ask: 'deny', plan: 'deny', auto: 'allow' },
  'kernel.environment.manage': { ask: 'deny', plan: 'deny', auto: 'os' },
}

/**
 * Projected tool catalogs by mode.
 * Ask: read, view, notebook (read only), kernel (status only), inspect, delegate (non-writing profiles).
 * Plan: Ask + tasks.
 * Auto: exactly ten tools.
 */
export const PIGENT_CATALOGS: Readonly<Record<PigentMode, readonly PigentToolName[]>> = {
  ask: ['read', 'view', 'notebook', 'kernel', 'inspect', 'delegate'],
  plan: ['read', 'view', 'notebook', 'kernel', 'inspect', 'delegate', 'tasks'],
  auto: PIGENT_TOOL_NAMES,
}

/** Per-mode action filters for multi-action tools. */
export const PIGENT_ACTION_FILTERS: Readonly<
  Record<'notebook' | 'kernel' | 'inspect' | 'tasks' | 'delegate', Readonly<Record<PigentMode, readonly string[]>>>
> = {
  notebook: {
    ask: ['read_cell'],
    plan: ['read_cell'],
    auto: ['read_cell', 'update_cell', 'insert_cell', 'delete_cell', 'move_cell', 'run_cell', 'add_markdown', 'clear_output'],
  },
  kernel: {
    ask: ['status', 'list_environments', 'operation_status'],
    plan: ['status', 'list_environments', 'operation_status'],
    auto: ['status', 'execute', 'interrupt', 'restart', 'shutdown', 'list_environments', 'operation_status', 'create_temporary', 'create_maintained', 'sync_environment', 'start_environment', 'promote_environment', 'delete_environment'],
  },
  inspect: {
    ask: ['variables', 'variable', 'dataframe', 'figure', 'object'],
    plan: ['variables', 'variable', 'dataframe', 'figure', 'object'],
    auto: ['variables', 'variable', 'dataframe', 'figure', 'object'],
  },
  tasks: {
    ask: [],
    plan: ['get', 'replace', 'patch'],
    auto: ['get', 'replace', 'patch'],
  },
  delegate: {
    ask: ['analysis', 'research', 'review'],
    plan: ['analysis', 'research', 'review'],
    auto: ['analysis', 'research', 'review', 'implementation'],
  },
}

// ---------------------------------------------------------------------------
// Result / error envelopes
// ---------------------------------------------------------------------------

export type Revision = `sha256:${string}`

export interface PigentToolError {
  code: PigentErrorCode
  message: string
  retryable: boolean
  details?: Record<string, unknown>
}

export interface PigentToolResult {
  version: 1
  ok: boolean
  summary: string
  data?: Record<string, unknown>
  artifacts?: ArtifactRef[]
  revisions?: { before: Revision; after: Revision } | null
  error?: PigentToolError
  warnings?: string[]
}

/** Trusted context injected by the Pigent host; the model can never override these fields. */
export interface PigentToolContext {
  protocol_version: typeof PIGENT_PROTOCOL_VERSION
  tool_call_id: string
  session_id: string
  workspace_id: string
  mode: PigentMode
  active_document?: { path: string; revision: Revision; cell_id: string }
  active_kernel_id?: string
}

// ---------------------------------------------------------------------------
// Session / tasks / terminal / artifact / interaction
// ---------------------------------------------------------------------------

export type OperationState = 'queued' | 'running' | 'waiting_for_user' | 'succeeded' | 'failed' | 'cancelled'
export type OperationOutcome = 'success' | 'partial' | 'failed' | 'cancelled' | 'superseded'
export type KernelEnvironmentKind = 'temporary' | 'maintained'
export type KernelEnvironmentStatus = 'provisioning' | 'ready' | 'stale' | 'syncing' | 'error' | 'deleting' | 'missing'

export interface OperationProgress {
  phase: string
  completed: number
  total?: number | null
  message?: string
}

export interface ToolReceipt {
  outcome: OperationOutcome
  summary: string
  identifiers?: Record<string, string>
  at: string
}

export interface OperationEnvelope {
  operation_id: string
  kind: string
  state: OperationState
  progress?: OperationProgress | null
  resource: { type: 'kernel_environment'; id: string }
  created_at: string
  updated_at: string
  session_id?: string | null
  tool_call_id?: string | null
  cancellable: boolean
  receipt?: ToolReceipt | null
  error?: PigentToolError | null
}

export interface KernelEnvironmentSummary {
  id: string
  kind: KernelEnvironmentKind
  name?: string | null
  display_name: string
  status: KernelEnvironmentStatus
  python_request: string
  python_version?: string | null
  interpreter?: string | null
  packages?: string[]
  source?: Record<string, unknown> | null
  lock_revision?: string | null
  revision: string
  created_at: string
  updated_at: string
  last_used_at?: string | null
  expires_at?: string | null
  active_kernel_ids?: string[]
  last_error?: PigentToolError | null
}

export interface ExecutionIdentity {
  username: string
  uid: number | string | null
  home: string
  workspace: string
}

export interface TaskNode {
  id: string
  title: string
  status: PigentTaskStatus
  depends_on?: string[]
  completion_criteria?: string[]
  children?: TaskNode[]
}

export interface TasksSnapshot {
  revision: string
  root: TaskNode
  updated_at?: string
}

export interface PigentSession {
  id: string
  account_id: string
  project_id: string
  workspace_id: string
  node_id: string
  mode: PigentMode
  approval_preference: 'automatic' | 'review_all'
  status: PigentSessionStatus
  title?: string
  created_at: string
  last_activity_at: string
  active_document?: { path: string; revision: Revision; cell_id: string }
  active_kernel_id?: string
  model?: { provider: string; model: string }
  tasks_snapshot?: TasksSnapshot
}

export interface TerminalSession {
  id: string
  name: string
  executable: string
  cwd: string
  status: 'running' | 'exited' | 'closed'
  cols: number
  rows: number
  created_at: string
  last_exit_code?: number | null
}

export interface ArtifactRef {
  id: string
  kind: PigentArtifactKind
  mime: string
  size: number
  created_at: string
  hash: Revision
  width?: number
  height?: number
  expires_at?: string
}

export interface PigentInteraction {
  version: 1
  interaction_id: string
  session_id: string
  tool_call_id: string
  kind: 'pty_handoff' | 'review_request' | 'clarification'
  summary: string
  shell_session_id?: string
  command_preview?: string
  choices: ('open_shell' | 'cancel' | 'allow_once' | 'allow_workspace' | 'reject')[]
}

export interface PigentEvent {
  version: 1
  event_id: number | null
  session_id: string
  type: PigentEventType
  timestamp: string
  payload?: Record<string, unknown>
}

// ---------------------------------------------------------------------------
// Runtime validation helpers
// ---------------------------------------------------------------------------

export function isPigentMode(value: unknown): value is PigentMode {
  return typeof value === 'string' && (PIGENT_MODES as readonly string[]).includes(value)
}

export function isPigentTool(value: unknown): value is PigentToolName {
  return typeof value === 'string' && (PIGENT_TOOL_NAMES as readonly string[]).includes(value)
}

export function isPigentEventType(value: unknown): value is PigentEventType {
  return typeof value === 'string' && (PIGENT_EVENT_TYPES as readonly string[]).includes(value)
}

export function isPigentErrorCode(value: unknown): value is PigentErrorCode {
  return typeof value === 'string' && (PIGENT_ERROR_CODES as readonly string[]).includes(value)
}

export function isRevision(value: unknown): value is Revision {
  return typeof value === 'string' && /^sha256:[0-9a-f]{64}$/.test(value)
}

export interface LegacySessionState {
  mode?: LegacyPigentMode
  requested_mode?: LegacyPigentMode
  effective_mode?: LegacyPigentMode
  [key: string]: unknown
}

/**
 * Legacy migration: mode/requested_mode/effective_mode == 'pilot' map to 'auto'
 * exactly once; legacy fields are removed after one successful write.
 * New schemas reject 'pilot' anywhere.
 */
export function migrateLegacySessionState(state: LegacySessionState): {
  migrated: Record<string, unknown>
  changed: boolean
} {
  const migrated: Record<string, unknown> = { ...state }
  let changed = false
  const map = (key: string): void => {
    if (key in migrated) {
      const value = migrated[key]
      if (value === 'pilot') {
        migrated[key] = 'auto'
        changed = true
      } else if (value !== 'ask' && value !== 'plan' && value !== 'auto') {
        throw new Error(`invalid legacy mode field ${key}: ${String(value)}`)
      }
    }
  }
  map('mode')
  map('requested_mode')
  map('effective_mode')
  if (changed) {
    delete migrated['requested_mode']
    delete migrated['effective_mode']
  }
  return { migrated, changed }
}

/** Tools advertised to the model in the selected mode. */
export function allowedToolNames(mode: PigentMode): readonly PigentToolName[] {
  return PIGENT_CATALOGS[mode]
}

/** Actions allowed for a multi-action tool in the selected mode. */
export function allowedActions(tool: PigentToolName, mode: PigentMode): readonly string[] {
  const filter = PIGENT_ACTION_FILTERS[tool as keyof typeof PIGENT_ACTION_FILTERS]
  if (!filter) return []
  return filter[mode]
}
